### ISSUE TITLE
Missing directories

### ISSUE THREAD

**Issue by kcmerrill** [karrick/godirwalk#1]
TL;DR
Disabling sorting fixes my issue, if it's sorted, directories go missing from the results.

Summary:
I cannot, for the life of me, figure out what's going on, but there is an issue where directories are not appearing properly when sort is enabled. 

I have a folder `/code` in my top level directory that when sort is enabled does not get returned in the results, but when sort is commented out or disabled via options it magically appears again. 

Looping over pre/post sort shows that the folder exists, but as soon as the next range takes place, it's just .... gone .... 

So here is what the code looks like(I just added some simple prints to walk.go)

walk.go:176

```

for _, deChild := range deChildren {
		if deChild.name == "code" {
			fmt.Println("debug:presort")
		}
	}
	if !options.Unsorted {
		sort.Sort(deChildren) // sort children entries unless upstream says to leave unsorted
	}

	for _, deChild := range deChildren {
		if deChild.name == "code" {
			fmt.Println("debug:postsort")
		}
	}

	for _, deChild := range deChildren {
		if deChild.name == "code" {
			fmt.Println("debug:innerrange")
		}
		osChildname := filepath.Join(osPathname, deChild.name)
```

```
09:02 PM ✘ kcmerrill  Desktop ] cat block.txt | grep debug
debug:presort
debug:postsort
09:02 PM ✔ kcmerrill  Desktop ]
```

I was expecting to see `debug:innerrange` as I wouldn't think simply sorting would make a difference. 

Ok, so now lets try again, but this time lets comment out `sort.Sort(deChildren)`.

```
for _, deChild := range deChildren {
		if deChild.name == "code" {
			fmt.Println("debug:presort")
		}
	}
	if !options.Unsorted {
		//sort.Sort(deChildren) // sort children entries unless upstream says to leave unsorted
	}

	for _, deChild := range deChildren {
		if deChild.name == "code" {
			fmt.Println("debug:postsort")
		}
	}

	for _, deChild := range deChildren {
		if deChild.name == "code" {
			fmt.Println("debug:innerrange")
		}
		osChildname := filepath.Join(osPathname, deChild.name)

```

And here are the results:

```
09:03 PM ✔ kcmerrill  Desktop ] cat block.txt | grep debug
debug:presort
debug:postsort
debug:innerrange
09:03 PM ✔ kcmerrill  Desktop ]
```

This is the strangest thing .... when doing a compare on the slices in regards to the lengths of them both, they are equal(as you can tell because both are visible in post sort). 

I'm on Mac(latest and greatest) and this is my go version:

```
09:07 PM ✔ kcmerrill  Desktop ] go version
go version go1.9.2 darwin/amd64
09:07 PM ✔ kcmerrill  Desktop ]
```

**Comment by karrick** [comment]
Could you please provide an output from `find $parentToCode -ls`? I'd like to re-create the hierarchy on my computer and try to reproduce.

Also, you mentioned you were on your Mac, and have the Go version. I assume you're using HPFS file system? Is the file system you're using case sensitive? If you're not sure, you can go to the Disk Utility app in the Utilities Folders in the Finder, then click on your your Apple hard drive, then click the info button, and look for "case sensitive" in the pop up window.

**Comment by karrick** [comment]
Hm, I went ahead and made a non-representative hierarchy in the `testdata` directory to see what might be the problem, and got the following results. First I modified the `walk.go` to have a mini-example of what you describe, but just a few files.

After adding the files, I went ahead and ran `go run examples/walk-fast/main.go testdata/dir6`, with the following results:

```
d--------- testdata/dir6
debug:presort
debug:postsort
d--------- testdata/dir6/abc
debug:inner
d--------- testdata/dir6/code
---------- testdata/dir6/code/123.txt
d--------- testdata/dir6/def
```

If would appear on my system that even with sorting enabled, as it is by default, it picks up `code`, and even enumerates into that directory as well. Could you please try running `go run examples/walk-fast/main.go` on your directory and see if it lists all the files? Could you also post your `Callback` function that you are using when you invoke `godirwalk.Walk`, including all of the functions that your callback function invokes. I *do* know that if the Callback function returns `filepath.SkipDir` on a directory it will stop processing that directory, and continue on to its siblings. If `filepath.SkipDir` is returned on a non-directory, the code still stop processing remaining siblings. This logic is there to imitate the behavior of `filepath.Walk`.

**Comment by karrick** [comment]
To illustrate what _might_ be the problem, but no guarantees because I do not have a copy of your filesystem on my disk... If I create another file in my hierarcy, let's say with the following command:

    date +%s > testdata/dir6/bravo.txt

And let's say I have the following simple change to my `Callback` function:

```Go
	err := godirwalk.Walk(dirname, &godirwalk.Options{
		Callback: func(osPathname string, de *godirwalk.Dirent) error {
			if de.Name() == "bravo.txt" {
				return filepath.SkipDir
			}
			fmt.Printf("%s %s\n", de.ModeType(), osPathname)
			return nil
		},
	})
```

What happens is that when the file hierarchy is enumerated with sorting on, the `bravo.txt` will be visited prior to `code`, and it just so happens that the above `Callback` function returns `filepath.SkipDir` error for a non-directory. In this case, `filepath.Walk` and `godirwalk.Walk` both stop processing all other siblings in this particular directory. If sorting is disabled, in my hypothetical scenario, and `code` is visited prior to `bravo.txt`, because that's the order the OS returned the directory entries in, then `code` would be visited, and then later `bravo.txt` would be visited and not printed.

So my hypothesis is that the `Callback` function is returning `filepath.SkipDir` when called for a non-directory file system entry in the same parent directory as the `code` directory, thus stopping processing for all additional directory entries, and the Callback will never be called with the `code` directory entry.

**Comment by kcmerrill** [comment]
My mac is using APFS and is not case sensitive. Having said that, let me make sure I'm _actually_ on the latest and greatest with my updates as that doesn't pass the sniff test for me.

So the directory(one of a few) that went "missing" in the results is located in the root `/code` folder so I wouldn't feel comfortable sending that over. However, here is the tree:

```
09:49 PM ✔ kcmerrill (master) block ] tree / -L 1
/
├── Applications
├── Library
├── Network
├── System
├── Users
├── Volumes
├── bin
├── code
├── cores
├── dev
├── etc -> private/etc
├── home
├── installer.failurerequests
├── net
├── private
├── sbin
├── tmp -> private/tmp
├── usr
└── var -> private/var

18 directories, 1 file
09:49 PM ✔ kcmerrill (master) block ]
```

Regarding the `filepath.SkipDir`, I thought of that too, except for the fact that it's literally 3 lines of code later its' just gone! Super weird ... let me make sure I have the latest updates to macosx and report back.

**Comment by kcmerrill** [comment]
Ohhh ..... That totally makes sense after re-reading what you mentioned! You're 100% right in regards to the `skipdir`. I think I was given a red herring with the sort due to how the folders were being returned. 

So, lets say I wanted to ignore a list of directories(in this case, I wanted to skip the `/Applications`  directory, how would you recommend that? Manually running a readdir, and then godirwalk? Or is there something else with skipdirectory that will continue on?

I don't think this is necessarily an issue now, just a little odd in it's behavior(it would stop reading the other directories on the same level)

**Comment by karrick** [comment]
So the behavior is odd, but it’s very useful, and I imagine there are very important use cases that neither you nor I have thought of, but are nevertheless important to account for.

In terms of your question about how to ignore directories, your Callback function merely returns `fillpath.SkipDir` for any directory you want to ignore. But only for *directories* you want to ignore. If you want to ignore a non directory then test the name at the beginning off the callback and return `nil`.

**Comment by kcmerrill** [comment]
Interesting. Ok, thanks for the clarification!
